domain/miningmanager/mempool/orphan_pool.go

```diff
@@ -0,0 +1,286 @@
+package mempool
+
+import (
+	"fmt"
+
+	"github.com/kaspanet/kaspad/domain/consensus/ruleerrors"
+
+	"github.com/kaspanet/kaspad/domain/consensus/utils/consensushashing"
+
+	"github.com/kaspanet/kaspad/domain/consensus/utils/utxo"
+
+	"github.com/kaspanet/kaspad/domain/consensus/model/externalapi"
+	"github.com/kaspanet/kaspad/domain/consensus/utils/estimatedsize"
+	"github.com/kaspanet/kaspad/domain/miningmanager/mempool/model"
+	"github.com/pkg/errors"
+)
+
+type idToOrphan map[externalapi.DomainTransactionID]*model.OrphanTransaction
+type previousOutpointToOrphans map[externalapi.DomainOutpoint]idToOrphan
+
+type orphansPool struct {
+	mempool                   *mempool
+	allOrphans                idToOrphan
+	orphansByPreviousOutpoint previousOutpointToOrphans
+	lastExpireScan            uint64
+}
+
+func newOrphansPool(mp *mempool) *orphansPool {
+	return &orphansPool{
+		mempool:                   mp,
+		allOrphans:                idToOrphan{},
+		orphansByPreviousOutpoint: previousOutpointToOrphans{},
+		lastExpireScan:            0,
+	}
+}
+
+// this function MUST be called with the mempool mutex locked for writes
+func (op *orphansPool) maybeAddOrphan(transaction *externalapi.DomainTransaction, isHighPriority bool) error {
+	serializedLength := estimatedsize.TransactionEstimatedSerializedSize(transaction)
+	if serializedLength > uint64(op.mempool.config.MaximumOrphanTransactionSize) {
+		str := fmt.Sprintf("orphan transaction size of %d bytes is "+
+			"larger than max allowed size of %d bytes",
+			serializedLength, op.mempool.config.MaximumOrphanTransactionSize)
+		return transactionRuleError(RejectBadOrphan, str)
+	}
+	if op.mempool.config.MaximumOrphanTransactionCount <= 0 {
```

Comment:
MaximumOrphanTransactionCount should probably never be below zero.

Reply:
But it can be zero.

Assuming we decide to leave it as int (for the reason listed above), what do we want to do if someone sets it to -1?
I assume we want the same behavior as when  it's 0, no?